drivers/virtio/virtio_mmio.c

```diff
@@ -0,0 +1,419 @@
+#define VIRTIO_MMIO_DEVICE_FEATURES_SEL     0x014
+#define VIRTIO_MMIO_DRIVER_FEATURES         0x020
+#define VIRTIO_MMIO_DRIVER_FEATURES_SEL     0x024
+#define VIRTIO_MMIO_QUEUE_SEL               0x030
+#define VIRTIO_MMIO_QUEUE_SIZE_MAX          0x034
+#define VIRTIO_MMIO_QUEUE_SIZE              0x038
+#define VIRTIO_MMIO_QUEUE_READY             0x044
+#define VIRTIO_MMIO_QUEUE_NOTIFY            0x050
+#define VIRTIO_MMIO_INTERRUPT_STATUS        0x060
+#define VIRTIO_MMIO_INTERRUPT_ACK           0x064
+#define VIRTIO_MMIO_STATUS                  0x070
+#define VIRTIO_MMIO_QUEUE_DESC_LOW          0x080
+#define VIRTIO_MMIO_QUEUE_DESC_HIGH         0x084
+#define VIRTIO_MMIO_QUEUE_DRIVER_LOW        0x090
+#define VIRTIO_MMIO_QUEUE_DRIVER_HIGH       0x094
+#define VIRTIO_MMIO_QUEUE_DEVICE_LOW        0x0a0
+#define VIRTIO_MMIO_QUEUE_DEVICE_HIGH       0x0a4
+#define VIRTIO_MMIO_QUEUE_SHM_SEL           0x0ac
+#define VIRTIO_MMIO_QUEUE_SHM_LEN_LOW       0x0b0
+#define VIRTIO_MMIO_QUEUE_SHM_LEN_HIGH      0x0b4
+#define VIRTIO_MMIO_QUEUE_SHM_BASE_LOW      0x0b8
+#define VIRTIO_MMIO_QUEUE_SHM_BASE_HIGH     0x0bc
+#define VIRTIO_MMIO_QUEUE_RESET             0x0c0
+#define VIRTIO_MMIO_QUEUE_CONFIG_GENERATION 0x0fc
+#define VIRTIO_MMIO_CONFIG                  0x100
+
+#define DEVICE_STATUS_ACKNOWLEDGE 0
+#define DEVICE_STATUS_DRIVER      1
+#define DEVICE_STATUS_DRIVER_OK   2
+#define DEVICE_STATUS_FEATURES_OK 3
+#define DEVICE_STATUS_NEEDS_RESET 6
+#define DEVICE_STATUS_FAILED      7
+
+#define VIRTIO_F_VERSION_1 32
+
+/* Ranges of feature bits for specific device types (see spec 2.2)*/
+#define DEV_TYPE_FEAT_RANGE_0_BEGIN 0
+#define DEV_TYPE_FEAT_RANGE_0_END   23
+#define DEV_TYPE_FEAT_RANGE_1_BEGIN 50
+#define DEV_TYPE_FEAT_RANGE_1_END   127
+
+#define VIRTIO_MMIO_QUEUE_INTERRUPT                1
+#define VIRTIO_MMIO_DEVICE_CONFIGURATION_INTERRUPT 2
+
+#define VIRTIO_MMIO_MAGIC             0x74726976
+#define VIRTIO_MMIO_SUPPORTED_VERSION 2
+#define VIRTIO_MMIO_INVALID_DEVICE_ID 0
+
+#define DEV_CFG(dev)  ((const struct virtio_mmio_config *)(dev)->config)
+#define DEV_DATA(dev) ((struct virtio_mmio_data *)(dev)->data)
+
+struct virtio_mmio_data {
+	DEVICE_MMIO_NAMED_RAM(reg_base);
+
+	struct virtq *virtqueues;
+	uint16_t virtqueue_count;
+
+	struct k_spinlock isr_lock;
+	struct k_spinlock notify_lock;
+};
+
+struct virtio_mmio_config {
+	DEVICE_MMIO_NAMED_ROM(reg_base);
+};
+
+static inline uint32_t virtio_mmio_read32(const struct device *dev, uint32_t offset)
+{
+	const mem_addr_t reg = DEVICE_MMIO_NAMED_GET(dev, reg_base) + offset;
+	uint32_t val;
+
+	barrier_dmem_fence_full();
+	val = sys_read32(reg);
+	return sys_le32_to_cpu(val);
+}
+
+static inline void virtio_mmio_write32(const struct device *dev, uint32_t offset, uint32_t val)
+{
+	const mem_addr_t reg = DEVICE_MMIO_NAMED_GET(dev, reg_base) + offset;
+
+	sys_write32(sys_cpu_to_le32(val), reg);
+	barrier_dmem_fence_full();
+}
+
+static void virtio_mmio_isr(const struct device *dev)
```

Comment:
The bulk of this symbol can be replaced with `virtio_isr(dev, isr_status, data->virtqueue_count)` from `virito_common.c`.

Reply:
Ditto.